src/wp-includes/html-api/class-wp-xml-tag-processor.php

```diff
@@ -123,6 +140,16 @@ class WP_XML_Tag_Processor {
 	 * precede or follow it.
 	 * 
 	 * See https://www.w3.org/TR/xml/#NT-document.
+	 * 
+	 * | Stage           | Meaning                                                             |
+	 * | ----------------|---------------------------------------------------------------------|
+	 * | *Prolog*        | The parser is parsing the prolog.                                   |
+	 * | *Element*       | The parser is parsing the root element.                             |
+	 * | *Misc*          | The parser is parsing miscellaneous content.                        |
```

Comment:
It might be helpful to start without an XML Tag Processor and go straight to the XML Processor, following the design of the HTML Processor with its `step()` and `step_in_body()` following the different _contexts_ of the HTML API spec.

_Prolog_, for example, sounds an awful lot like the `IN PROLOG` context, where the rules can be contained and indicate a state transition rather than embedding the rules for staging in the lower-level tag processing.

Reply:
Yup, good call

Reply:
> Shipping only an XML Processor was quite tedious

out of curiosity, what was tedious about it? I have a hard time understanding the need for an XML tag processor, since the nesting rules are straightforward when compared with HTML, where the complicated semantic rules forced a split

Reply:
I found it difficult to:

- Think about two different levels of parsing (tokenization, context) in a single class
- Avoid intertwining them
- Come up with names for all the methods


That being said, there's nothing inherently wrong with the approach and I can see how it could work.